### PR TITLE
Add tests for APER encoding/decoding

### PR DESCRIPTION
**Pull request by OYTIS**
Adding tests is not a big deal by itself, but it seems there are still a number of small issues in APER encoders/decoders. The very first test

```
<PDU>
	<str-m></str-m>
</PDU>
```

seems to produce an incorrect encoding for me (I've checked against http://asn1-playground.oss.com/). Maybe we could leave the tests as a PR and fix the issues one-by-one. Or maybe there is an obvious error in the test procedure itself you'll be able to identify.

### CODE REVIEW

File: tests/tests-c-compiler/check-src/check-126.-gen-PER.c [hunk 1]
```diff
@@ -65,7 +66,12 @@ save_object_as(PDU_t *st, enum enctype how) {
 	 * Save object using specified method.
 	 */
 	switch(how) {
-	case AS_PER:
+	case AS_APER:
```
Comment (mouse07410) [inline tests/tests-c-compiler/check-src/check-126.-gen-PER.c]: I think that `AS_PER` should be the same as `AS_UPER`, **not** as `AS_APER`. This case selector should be moved down to before line 74, I think.

**Comment by mouse07410** [comment]
First, there's this difference in the test output:
```
=== original ===
<PDU>
        <str-m></str-m>
</PDU>
=== re-encoded ===
<PDU>
    <str-m></str-m>
</PDU>
=== end ===
```
The original indents `<str-m></str-m>` by a tab (8 spaces), and the re-encoded indents by blanks (4 spaces).

Then, there's this error:

```
Processing file [../data-126-09-C.in]
APER encoding
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 1
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 2
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 3
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 4
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 5
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 6
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 7
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 8
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 9
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 10
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 11
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 12
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 13
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 14
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 15
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 16
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 17
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 18
LOADING OBJECT OF SIZE 30 FROM [data-126-09-C.in] as XER, chunks 19
PDU ::= {
    str-m:
}
SAVED OBJECT IN SIZE 4
Comparing PER output with [.././data-126/data-126-09-C.out]
Trying to decode [.././data-126/data-126-09-C.out]
LOADING OBJECT OF SIZE 4 FROM [.././data-126/data-126-09-C.out] as PER, chunks 1
-> PER decode error (0 bits of 4 bytes (c=4,l=4))
Assertion failed: (rval.code == RC_OK), function load_object_from, file check-126.-gen-PER.c, line 207.
make[6]: *** [check-succeeded] Abort trap: 6
FAIL check-src/check-126.-gen-PER.c (exit status: 2)
```

**Comment by OYTIS** [comment]
Sorry, I don't get it. As far as I understand, APER decoder can't read what APER encoder has written. How can it be affected by indentation in XER input?

**Comment by velichkov** [comment]
Hi @OYTIS,

> but it seems there are still a number of small issues in APER encoders/decoders. 

Yes there are quite a lot of issues with the APER encoding.

> Maybe we could leave the tests as a PR and fix the issues one-by-one.

This is a good approach but personally I prefer the other approach to add a fix together with the tests that reproduce the problem in a single PR, and I prefer small PRs that are easier to review and test.

> Or maybe there is an obvious error in the test procedure itself you'll be able to identify.

My initial idea was not to add the APER tests in `check-126.-gen-PER.c` (although this is a valid approach) but to use `check-126.-gen-PER.c` as an example and write a new tests specifically for APER encoding.

>  As far as I understand, APER decoder can't read what APER encoder has written. 

It doesn't decode what it has previously encoded but instead it tries to decode the content of a binary file that is in the repo.

> Trying to decode [.././data-126/data-126-09-C.out]

The obvious error is that the `data-126/data-126-09-C.out` file contains an `UPER` encoded message so it obviously :smile: can't be decoded as `APER`. You need to generate an `APER` encoded binaries (with a different extension because you don't want to overwrite the UPER ones)

https://github.com/mouse07410/asn1c/blob/102c97ce50b7341a24c9d0d06bcf196afaf258d9/tests/tests-c-compiler/check-src/check-126.-gen-PER.c#L266-L269

To regenerate all files you need to fix the above if clause (remove `compare`), change the extension  and run `REGENERATE=1 ./check-program` (it currently fails when compiled with ASAN)

See also https://github.com/mouse07410/asn1c/blob/102c97ce50b7341a24c9d0d06bcf196afaf258d9/tests/tests-c-compiler/data-126/README#L1-L11

**Comment by OYTIS** [comment]
@velichkov Thank you. I'll just close this PR then and rewrite according to your suggestions.

**Comment by mouse07410** [comment]
@OYTIS any luck on updated tests so far?

**Pull request closed without merging**